go/adbc/driver/bigquery/record_reader.go

```diff
@@ -0,0 +1,315 @@
+	includeName := parameterMode == OptionValueQueryParameterModeNamed
+	schema := values.Schema()
+	for i, v := range values.Columns() {
+		pi, err := arrowValueToQueryParameterValue(schema.Field(i), v, row)
+		if err != nil {
+			return nil, err
+		}
+		parameters[i] = pi
+		if includeName {
+			parameters[i].Name = values.ColumnName(i)
+		}
+	}
+	return parameters, nil
+}
+
+func runPlainQuery(ctx context.Context, query *bigquery.Query, alloc memory.Allocator, resultRecordBufferSize int) (bigqueryRdr *reader, totalRows int64, err error) {
+	arrowIterator, totalRows, err := runQuery(ctx, query, false)
+	if err != nil {
+		return nil, -1, err
+	}
+	rdr, err := ipcReaderFromArrowIterator(arrowIterator, alloc)
+	if err != nil {
+		return nil, -1, err
+	}
+
+	chs := make([]chan arrow.Record, 1)
+	ctx, cancelFn := context.WithCancel(ctx)
+	ch := make(chan arrow.Record, resultRecordBufferSize)
+	chs[0] = ch
+
+	defer func() {
+		if err != nil {
+			close(ch)
+			cancelFn()
+		}
+	}()
+
+	bigqueryRdr = &reader{
+		refCount:   1,
+		chs:        chs,
+		curChIndex: 0,
+		err:        nil,
+		cancelFn:   cancelFn,
+		schema:     nil,
+	}
+
+	go func() {
+		defer rdr.Release()
+		for rdr.Next() && ctx.Err() == nil {
+			rec := rdr.Record()
+			rec.Retain()
+			ch <- rec
+		}
+
+		err = checkContext(ctx, rdr.Err())
+		defer close(ch)
+	}()
+	return bigqueryRdr, totalRows, nil
+}
+
+// kicks off a goroutine for each endpoint and returns a reader which
+// gathers all of the records as they come in.
+func newRecordReader(ctx context.Context, query *bigquery.Query, boundParameters array.RecordReader, parameterMode string, alloc memory.Allocator, resultRecordBufferSize, prefetchConcurrency int) (bigqueryRdr *reader, totalRows int64, err error) {
+	if boundParameters == nil {
+		return runPlainQuery(ctx, query, alloc, resultRecordBufferSize)
+	}
+
+	recs := make([]arrow.Record, 0)
+	for boundParameters.Next() {
+		rec := boundParameters.Record()
+		recs = append(recs, rec)
+	}
+	batches := int64(len(recs))
+	chs := make([]chan arrow.Record, batches)
+
+	ch := make(chan arrow.Record, resultRecordBufferSize)
+	group, ctx := errgroup.WithContext(ctx)
+	group.SetLimit(prefetchConcurrency)
+	ctx, cancelFn := context.WithCancel(ctx)
+	chs[0] = ch
+
+	defer func() {
+		if err != nil {
+			close(ch)
+			cancelFn()
+		}
+	}()
+
+	bigqueryRdr = &reader{
+		refCount: 1,
+		chs:      chs,
+		err:      nil,
+		cancelFn: cancelFn,
+		schema:   nil,
+	}
+
+	rec := recs[0]
+	for i := 0; i < int(rec.NumRows()); i++ {
+		parameters, err := getQueryParameter(rec, i, parameterMode)
+		if err != nil {
+			return nil, -1, err
+		}
+		if parameters != nil {
+			query.QueryConfig.Parameters = parameters
+		}
+
+		arrowIterator, rows, err := runQuery(ctx, query, false)
+		if err != nil {
+			return nil, -1, err
+		}
+		totalRows = rows
+		rdr, err := ipcReaderFromArrowIterator(arrowIterator, alloc)
+		if err != nil {
+			return nil, -1, err
+		}
```

Comment:
this portion should probably be a helper function since it's repeated below, you also need to call `Release()` on `rec` after this loop is completed